File: python_kata/src/data_structures/ring_buffer.py
from array import array, ArrayType
class RingBuffer:
    __head: int
    __tail: int
    __capacity: int
    __length: int
    __array: ArrayType[int]

    def __init__(self, __capacity: int = 0):
        self.__head = self.__tail = 0
        self.__capacity = __capacity
        self.__length = 0
        arr = [0 for _ in range(self.__capacity)]
        self.__array = array("I", arr)

        # Free memory
        arr = None

    def __copyArray(self) -> None:
        newArr = [0 for _ in range(self.__capacity * 2)]
        length = 0
        for i in range(self.__head, self.__capacity, 1):
            newArr[i] = self.__array[i]
            length += 1
        for j in range(self.__tail):
            newArr[length] = self.__array[j]
            length += 1

        self.__head = 0
        self.__tail = length
        self.__length = length
        self.__capacity = self.__capacity * 2 
        self.__array = array("I", newArr)

    def push(self, item: int) -> None:
        if self.__tail == self.__head and self.__length != 0:
            self.__copyArray()
        self.__array[self.__tail] = item
        self.__length += 1

        self.__tail = (self.__tail + 1) % self.__capacity

        return None

    def pop(self) -> int | None:
        if self.__length == 0:
            return None
        item = None
        if self.__tail - 1 < 0:
            self.__tail = self.__capacity - 1
            item = self.__array[self.__tail]
        else:
            self.__tail -= 1
            item = self.__array[self.__tail]

        self.__length -= 1

        return item

    def unshift(self, item: int) -> None:
        if self.__length == 0:
            self.push(item)

            return None

        if self.__head == self.__tail and self.__length != 0:
            self.__copyArray()

        if self.__head == 0:
            self.__head = self.__capacity - 1
        else:
            self.__head -= 1

        self.__array[self.__head] = item
        self.__length += 1

        return None

    def shift(self) -> int | None:
        if self.__length == 0:
            return None
        item = self.__array[self.__head]
        self.__head += 1
        if self.__head >= self.__capacity:
            self.__head = self.__head % self.__capacity

        self.__length -= 1
        return item

    def get(self, indx: int) -> int | None:
        if indx >= self.__length or indx < -self.__length:
            return None

        if indx < 0:
            if self.__tail + indx >= 0:
                return self.__array[self.__tail + indx]
            else:
                i = self.__capacity + (self.__tail + indx)
                return self.__array[i]
        
        trueIndex = (self.__head + indx) % self.__capacity
        if (trueIndex < self.__head and trueIndex >= self.__tail):
            return None

        return self.__array[trueIndex]


    def getLength(self) -> int:
        return self.__length

    def getCapacity(self) -> int:
        return self.__capacity

    def getHead(self) -> int:
        return self.__head

    def getTail(self) -> int:
        return self.__tail

File: typescript-kata/src/data-structures/ring-buffer.ts
/*
This is simplified implementation of the ring bufer (aka. circular buffer).
If you want to create better implementation think about underlaying data
structure that can accept multiple types of values.
*/

export class RingBuffer {
    private head: number;
    private tail: number;
    private capacity: number;
    private length: number;
    private array: Uint8Array;

    constructor(capacity: number = 0) {
        this.capacity = capacity;
        this.head = this.tail = 0;
        this.length = 0
        this.array = new Uint8Array(this.capacity);
    }

    private copyArray(): void {
        const newArray = new Uint8Array(this.capacity * 2);

        let length = 0;
        for (let i = this.head; i < this.capacity; i++, length++){
            newArray[length] = this.array[i]
        }
        for (let k = 0; k < this.tail; ++k, ++length) {
            newArray[length] = this.array[k]
        }
        this.head = 0
        this.tail = length
        this.capacity *= 2
        this.length = length
        this.array = newArray

    }

    push(item: number): void {
        if (this.tail === this.head && this.length !== 0) {
            this.copyArray()
        }
        this.array[this.tail] = item
        this.length++

        this.tail = (this.tail + 1) % this.capacity

    }

    pop(): number | undefined {
        if (this.length === 0) {
            return undefined
        }

        let item;
        if (this.tail === 0) {
            this.tail = this.capacity - 1
            item = this.array[this.tail]
        } else {
            this.tail--
            item = this.array[this.tail]
        }
        this.length--

        return item
    }

    unshift(item: number): void {
        if (this.length === 0) {
            this.push(item)
            return
        }

        if (this.head === this.tail && this.length !== 0) {
            this.copyArray()
        }
        
        if (this.head - 1 < 0) {
            this.head = this.capacity - 1
        } else {
            this.head--
        }
        this.array[this.head] = item
        this.length++
        return
    }

    shift(): number | undefined {
        if (this.length === 0) {
            return undefined
        }
        const item = this.array[this.head]
        this.head++
        if (this.head >= this.capacity) {
            this.head = this.head % this.capacity
        }
        this.length--

        return item
    }

    get(indx: number): number | undefined {
        if (this.length === 0) {
            return undefined
        }
        if (indx >= this.length || indx < -this.length) {
            return undefined
        }
        if (indx < 0) {
            if (this.tail + indx >= 0) {
                return this.array[this.tail + indx]
            } else {
                const i = this.capacity + (this.tail + indx)
                return this.array[i]
            }
        }
        const trueIndex = (this.head + indx) % this.capacity
        if (trueIndex < this.head && trueIndex >= this.tail) {
            return undefined
        }

        return this.array[trueIndex]
    }

    getLength(): number {
        return this.length
    }

    getCapacity(): number {
        return this.capacity
    }

    getHead(): number {
        return this.head
    }

    getTail(): number {
        return this.tail
    }
}
